pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package rabbitmq
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+// SubjectAllChannels represents subject to subscribe for all the channels.
+const (
+	chansPrefix        = "channels"
+	subjectAllChannels = "channels.>"
+	routingKey         = "mainfluxkey"
+	exchange           = "mainflux"
+	exchangeKind       = "fanout"
+	queueDurability    = true
+	queueDelete        = false
+	queueExclusivity   = false
+	queueWait          = false
+	consumerTag        = "mainflux-consumer"
+	mandatory          = false
+	immediate          = false
+)
+
+var (
+	errAlreadySubscribed = errors.New("already subscribed to topic")
+	errNotSubscribed     = errors.New("not subscribed")
+	errEmptyTopic        = errors.New("empty topic")
+)
+
+var _ messaging.PubSub = (*pubsub)(nil)
+
+// PubSub wraps messaging Publisher exposing
+// Close() method for RabbitMQ connection.
+type PubSub interface {
+	messaging.PubSub
+	Close()
+}
+
+type pubsub struct {
+	conn          *amqp.Connection
+	logger        log.Logger
+	queue         amqp.Queue
+	channel       *amqp.Channel
+	subscriptions map[string]bool
+	done          chan error
+	mutex         sync.Mutex
+}
+
+// NewPubSub returns RabbitMQ message publisher/subscriber.
+func NewPubSub(url, queueName string, logger log.Logger) (PubSub, error) {
+	endpoint := fmt.Sprintf("amqp://%s", url)
+	conn, err := amqp.Dial(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, err
+	}
+	queue, err := ch.QueueDeclare(queueName, queueDurability, queueDelete, queueExclusivity, queueWait, nil)
+	if err != nil {
+		return nil, err
+	}
+	ret := &pubsub{
+		conn:          conn,
+		queue:         queue,
+		channel:       ch,
+		logger:        logger,
+		subscriptions: make(map[string]bool),
+		done:          make(chan error),
+	}
+	return ret, nil
+}
+
+func (ps *pubsub) Publish(topic string, msg messaging.Message) error {
```

Comment:
There is no need to reimplement this in `pubsub`, you can use the implementation from the `publisher.go`:
```go

type pubsub struct {
	publisher
	logger log.Logger
	queue  amqp.Queue
	subscriptions map[string]bool
	mutex sync.Mutex
}
```

Reply:
Good idea

Reply:
I am getting this error `panic: runtime error: invalid memory address or nil pointer dereferenc`. I have resorted to another approach